src/common/ina.rs

```diff
@@ -0,0 +1,246 @@
+    Ok(Arc::new(Mutex::new(from_config(Model::Ina226, cfg, deps)?)))
+}
+
+fn from_config(
+    model: Model,
+    cfg: ConfigType,
+    dependencies: Vec<Dependency>,
+) -> anyhow::Result<Ina<I2cHandleType>> {
+    let i2c_address = cfg
+        .get_attribute::<u8>("i2c_address")
+        .unwrap_or(DEFAULT_I2C_ADDRESS);
+
+    let default_max_current_amperes = match &model {
+        Model::Ina219 => 3.2,
+        Model::Ina226 => 20.0,
+    };
+    let max_current_amperes = cfg
+        .get_attribute::<f64>("max_current_amps")
+        .unwrap_or(default_max_current_amperes);
+    let max_current_nano_amperes = (max_current_amperes * 1e9) as i64;
+
+    let shunt_resistance_ohms = cfg
+        .get_attribute::<f64>("shunt_resistance")
+        .unwrap_or(DEFAULT_SHUNT_RESISTANCE_OHMS);
+    let shunt_resistance_nano_ohms = (shunt_resistance_ohms * 1e9) as i64;
+
+    let i2c_name = cfg
+        .get_attribute::<String>("i2c_bus")
+        .map_err(|_| anyhow::anyhow!("i2c_bus is a required config attribute for {}", model))?;
+    let board = get_board_from_dependencies(dependencies).ok_or(anyhow::anyhow!(
+        "a board must be configured to configure an {} sensor",
+        model
+    ))?;
+    let i2c_handle = board.get_i2c_by_name(i2c_name)?;
+
+    Ina::new(
+        model,
+        i2c_handle,
+        i2c_address,
+        max_current_nano_amperes,
+        shunt_resistance_nano_ohms,
+    )
+}
+
+#[derive(Clone, Copy)]
+pub enum Model {
+    Ina219,
+    Ina226,
+}
+
+impl fmt::Display for Model {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let model_str = match self {
+            Model::Ina219 => "INA219",
+            Model::Ina226 => "INA226",
+        };
+        write!(f, "{model_str}")
+    }
+}
+
+pub struct Ina<H> {
+    model: Model,
+    i2c_handle: H,
+    i2c_address: u8,
+    calibration_scale_bytes: [u8; 2],
+    max_current_nano_amperes: i64,
+    power_reading_lsb: i64,
+}
+
+impl<H> Ina<H>
+where
+    H: I2CHandle,
+{
+    fn new(
+        model: Model,
+        i2c_handle: H,
+        i2c_address: u8,
+        max_current_nano_amperes: i64,
+        shunt_resistance_nano_ohms: i64,
+    ) -> anyhow::Result<Self> {
+        let current_reading_lsb = max_current_nano_amperes / (1 << 15);
+        let (unadjusted_calibration_scale, power_reading_lsb) = match model {
+            Model::Ina219 => {
+                let power_lsb = (max_current_nano_amperes * 20 + (1 << 14)) / (1 << 15);
+                (INA_219_CALIBRATION_SCALE, power_lsb)
+            }
+            Model::Ina226 => (INA_226_CALIBRATION_SCALE, 25 * current_reading_lsb),
+        };
+        let calibration_scale = (unadjusted_calibration_scale
+            / ((current_reading_lsb as f64) * (shunt_resistance_nano_ohms as f64 * 1e-9)))
+            as i64;
+        if calibration_scale >= (1 << 16) {
+            anyhow::bail!(
+                "ina calibrate: calibration register value invalid {:?}",
```

Comment:
Maybe `"ina calibrate: invalid calibration scale value {:?} exceeds limit 1 << 16"`. I'm not sure though, since it is a computed value, so it isn't immediately clear how the user should remedy it.